{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "source": [
    "# Example (synthetic) data sets\n",
    "For the purpose of this effort to create a common data shape on Glioblastoma, two synthetic data sets were provided. Synthetic datasets contain fabricated data that is represtable of actual data on the topic. \n",
    "\n",
    "In this chapter we will inspect the provided datasets and prepare them to by converted into linked data. As described earlier, linked data is data that follows a linked-data format consisting of triples where the individual parts consists of either values or identifiers (IRIs) pointing to concept definitions. \n",
    "\n",
    "The proces of converting data to linked data is rather straightforwards and can be summed up as follows:\n",
    "1. Inspect the data\n",
    "2. Clean the data headers\n",
    "3. Align the data with common controlled vocabularies or ontologies\n",
    "4. Design linked-data shapes using the IRIs from the controlled vocabularies and ontologies\n",
    "5. Transform the data into RDF\n",
    "\n",
    "In this chapter we will focus on the first two steps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataset 1 on Glioblastoma\n",
    "The first dataset is a provided excel file (Sample clinical data_July 2022.xlsx) ## TODO add a link.\n",
    "It contains synthetic clinical data on patients with glioblastoma. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID Lab site</th>\n",
       "      <th>ID clinical site</th>\n",
       "      <th>DOB</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Pathology</th>\n",
       "      <th>Localization</th>\n",
       "      <th>IDH1/2 status</th>\n",
       "      <th>MGMT status</th>\n",
       "      <th>ECOG</th>\n",
       "      <th>...</th>\n",
       "      <th>doses bevacizumab</th>\n",
       "      <th>Best response</th>\n",
       "      <th>pseudo-progression (1=yes 0=no)</th>\n",
       "      <th>œdema</th>\n",
       "      <th>PP or œdema</th>\n",
       "      <th>progression (1=yes 0=no)</th>\n",
       "      <th>date of progression</th>\n",
       "      <th>months PFS</th>\n",
       "      <th>deceased (1=yes 0=no)</th>\n",
       "      <th>months OS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ge 829</td>\n",
       "      <td>DAG</td>\n",
       "      <td>1953-05-16</td>\n",
       "      <td>62</td>\n",
       "      <td>F</td>\n",
       "      <td>GBM</td>\n",
       "      <td>parietal R</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>PD</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2013-11-11</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ge 835</td>\n",
       "      <td>EAG</td>\n",
       "      <td>1964-06-18</td>\n",
       "      <td>50</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>temporal D</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2014-01-21</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ge 849</td>\n",
       "      <td>FAG</td>\n",
       "      <td>1957-12-15</td>\n",
       "      <td>57</td>\n",
       "      <td>F</td>\n",
       "      <td>GBM</td>\n",
       "      <td>frontal R</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2014-09-02</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ge 852</td>\n",
       "      <td>GAG</td>\n",
       "      <td>1960-03-22</td>\n",
       "      <td>55</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>fronto insular L</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>SD</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2014-10-01</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ge 882</td>\n",
       "      <td>IAG</td>\n",
       "      <td>1948-05-06</td>\n",
       "      <td>67</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>fronto insular R</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>PD</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2014-12-22</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Ge 893*</td>\n",
       "      <td>KAG</td>\n",
       "      <td>1958-07-12</td>\n",
       "      <td>56</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>temporal R</td>\n",
       "      <td>wt</td>\n",
       "      <td>Meth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>SD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2015-02-04</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ge 901</td>\n",
       "      <td>LAG</td>\n",
       "      <td>1991-06-09</td>\n",
       "      <td>23</td>\n",
       "      <td>F</td>\n",
       "      <td>AIII</td>\n",
       "      <td>fronto temporal insular R</td>\n",
       "      <td>mut</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>29</td>\n",
       "      <td>SD</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2016-07-18</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Ge 904</td>\n",
       "      <td>MAG</td>\n",
       "      <td>1949-10-13</td>\n",
       "      <td>65</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>parieto occipital L</td>\n",
       "      <td>wt</td>\n",
       "      <td>Meth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>PR</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2016-12-12</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Ge 939</td>\n",
       "      <td>PAG</td>\n",
       "      <td>1972-04-10</td>\n",
       "      <td>43</td>\n",
       "      <td>F</td>\n",
       "      <td>AIII</td>\n",
       "      <td>temporal R</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>42</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2015-06-12</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Ge 941*</td>\n",
       "      <td>OAG</td>\n",
       "      <td>1954-01-07</td>\n",
       "      <td>61</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>frontal L</td>\n",
       "      <td>mut</td>\n",
       "      <td>Meth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>CR</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaT</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Ge 950</td>\n",
       "      <td>QAG</td>\n",
       "      <td>1947-03-22</td>\n",
       "      <td>68</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>parietal L</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>PD</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2015-09-25</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Ge 963*</td>\n",
       "      <td>RAG</td>\n",
       "      <td>1941-06-14</td>\n",
       "      <td>73</td>\n",
       "      <td>M</td>\n",
       "      <td>AIII</td>\n",
       "      <td>frontal L</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>SD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2017-03-07</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Ge 970</td>\n",
       "      <td>UAG</td>\n",
       "      <td>1971-05-01</td>\n",
       "      <td>44</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>pareito occipital R</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>PD</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2016-01-13</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Ge 977</td>\n",
       "      <td>VAG</td>\n",
       "      <td>1947-06-28</td>\n",
       "      <td>68</td>\n",
       "      <td>F</td>\n",
       "      <td>GBM</td>\n",
       "      <td>temporal L</td>\n",
       "      <td>wt</td>\n",
       "      <td>Meth</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>SD</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2017-03-03</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Ge 989*</td>\n",
       "      <td>WAG</td>\n",
       "      <td>1974-11-10</td>\n",
       "      <td>41</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>temporal L</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>PD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2015-11-10</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Ge 993</td>\n",
       "      <td>XAG</td>\n",
       "      <td>1951-12-07</td>\n",
       "      <td>63</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>frontal L precentral</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>PR</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2016-06-14</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ge 1006</td>\n",
       "      <td>YAG</td>\n",
       "      <td>1955-11-05</td>\n",
       "      <td>60</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>frontotemporal L</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>PD</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2016-07-29</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Ge 1007</td>\n",
       "      <td>ZAG</td>\n",
       "      <td>1942-05-15</td>\n",
       "      <td>73</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>frontal L</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>PD</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2016-06-22</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Ge 1018</td>\n",
       "      <td>ABG</td>\n",
       "      <td>1975-07-11</td>\n",
       "      <td>41</td>\n",
       "      <td>M</td>\n",
       "      <td>GBM</td>\n",
       "      <td>parieto temporal R</td>\n",
       "      <td>wt</td>\n",
       "      <td>UnMeth</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>PD</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2016-11-10</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID Lab site ID clinical site        DOB  Age  Gender Pathology  \\\n",
       "0       Ge 829              DAG 1953-05-16    62      F       GBM   \n",
       "1       Ge 835              EAG 1964-06-18    50      M       GBM   \n",
       "2       Ge 849              FAG 1957-12-15    57      F       GBM   \n",
       "3       Ge 852              GAG 1960-03-22    55      M       GBM   \n",
       "4       Ge 882              IAG 1948-05-06    67      M       GBM   \n",
       "5      Ge 893*              KAG 1958-07-12    56      M       GBM   \n",
       "6       Ge 901              LAG 1991-06-09    23      F      AIII   \n",
       "7       Ge 904              MAG 1949-10-13    65      M       GBM   \n",
       "8       Ge 939              PAG 1972-04-10    43      F      AIII   \n",
       "9      Ge 941*              OAG 1954-01-07    61      M       GBM   \n",
       "10      Ge 950              QAG 1947-03-22    68      M       GBM   \n",
       "11     Ge 963*              RAG 1941-06-14    73      M      AIII   \n",
       "12      Ge 970              UAG 1971-05-01    44      M       GBM   \n",
       "13      Ge 977              VAG 1947-06-28    68      F       GBM   \n",
       "14     Ge 989*              WAG 1974-11-10    41      M       GBM   \n",
       "15      Ge 993              XAG 1951-12-07    63      M       GBM   \n",
       "16     Ge 1006              YAG 1955-11-05    60      M       GBM   \n",
       "17     Ge 1007              ZAG 1942-05-15    73      M       GBM   \n",
       "18     Ge 1018              ABG 1975-07-11    41      M       GBM   \n",
       "\n",
       "                Localization  IDH1/2 status MGMT status  ECOG  ...  \\\n",
       "0                  parietal R            wt      UnMeth     1  ...   \n",
       "1                  temporal D            wt      UnMeth     0  ...   \n",
       "2                   frontal R            wt      UnMeth     0  ...   \n",
       "3            fronto insular L            wt      UnMeth     0  ...   \n",
       "4            fronto insular R            wt      UnMeth     0  ...   \n",
       "5                  temporal R            wt        Meth     0  ...   \n",
       "6   fronto temporal insular R           mut      UnMeth     0  ...   \n",
       "7         parieto occipital L            wt        Meth     0  ...   \n",
       "8                  temporal R            wt      UnMeth     1  ...   \n",
       "9                   frontal L           mut        Meth     0  ...   \n",
       "10                 parietal L            wt      UnMeth     1  ...   \n",
       "11                  frontal L            wt      UnMeth     1  ...   \n",
       "12        pareito occipital R            wt      UnMeth     0  ...   \n",
       "13                 temporal L            wt        Meth     1  ...   \n",
       "14                temporal L             wt      UnMeth     0  ...   \n",
       "15       frontal L precentral            wt      UnMeth     1  ...   \n",
       "16           frontotemporal L            wt      UnMeth     0  ...   \n",
       "17                  frontal L            wt      UnMeth     1  ...   \n",
       "18         parieto temporal R            wt      UnMeth     0  ...   \n",
       "\n",
       "   doses bevacizumab  Best response  pseudo-progression (1=yes 0=no) œdema  \\\n",
       "0                 12             PD                                0     1   \n",
       "1                  0             PD                                1     1   \n",
       "2                  9             PD                                1     0   \n",
       "3                 16             SD                                0     1   \n",
       "4                  4             PD                                0     1   \n",
       "5                  0             SD                                0     0   \n",
       "6                 29             SD                                1     0   \n",
       "7                  6             PR                                1     1   \n",
       "8                 42             PD                                1     1   \n",
       "9                  0             CR                                1     0   \n",
       "10                 8             PD                                0     1   \n",
       "11                 0             SD                                0     0   \n",
       "12                 8             PD                                0     1   \n",
       "13                 0             SD                                1     0   \n",
       "14                23             PD                                0     0   \n",
       "15                 7             PR                                1     0   \n",
       "16                14             PD                                0     1   \n",
       "17                14             PD                                0     0   \n",
       "18                15             PD                                0     1   \n",
       "\n",
       "    PP or œdema  progression (1=yes 0=no)  date of progression months PFS  \\\n",
       "0             1                         1           2013-11-11          3   \n",
       "1             1                         1           2014-01-21          6   \n",
       "2             1                         1           2014-09-02         11   \n",
       "3             1                         1           2014-10-01         11   \n",
       "4             1                         1           2014-12-22         18   \n",
       "5             0                         1           2015-02-04         26   \n",
       "6             1                         1           2016-07-18         17   \n",
       "7             1                         1           2016-12-12         29   \n",
       "8             1                         1           2015-06-12          6   \n",
       "9             1                         0                  NaT         38   \n",
       "10            1                         1           2015-09-25          7   \n",
       "11            0                         1           2017-03-07         19   \n",
       "12            1                         1           2016-01-13          6   \n",
       "13            1                         1           2017-03-03         20   \n",
       "14            0                         1           2015-11-10          5   \n",
       "15            1                         1           2016-06-14         12   \n",
       "16            1                         1           2016-07-29         10   \n",
       "17            0                         1           2016-06-22          7   \n",
       "18            1                         1           2016-11-10         10   \n",
       "\n",
       "    deceased (1=yes 0=no)  months OS  \n",
       "0                       1         19  \n",
       "1                       1         10  \n",
       "2                       1         18  \n",
       "3                       1         21  \n",
       "4                       1         24  \n",
       "5                       0         44  \n",
       "6                       1         37  \n",
       "7                       1         35  \n",
       "8                       1         32  \n",
       "9                       0         38  \n",
       "10                      1         11  \n",
       "11                      0         33  \n",
       "12                      1         13  \n",
       "13                      1         29  \n",
       "14                      0         30  \n",
       "15                      1         18  \n",
       "16                      1         18  \n",
       "17                      1         15  \n",
       "18                      1         18  \n",
       "\n",
       "[19 rows x 26 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "gdb1 = pd.read_excel(\"data/Sample clinical data_July 2022.xlsx\")\n",
    "gdb1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 19 entries, 0 to 18\n",
      "Data columns (total 26 columns):\n",
      " #   Column                                    Non-Null Count  Dtype         \n",
      "---  ------                                    --------------  -----         \n",
      " 0   ID Lab site                               19 non-null     object        \n",
      " 1   ID clinical site                          19 non-null     object        \n",
      " 2   DOB                                       19 non-null     datetime64[ns]\n",
      " 3   Age                                       19 non-null     int64         \n",
      " 4   Gender                                    19 non-null     object        \n",
      " 5   Pathology                                 19 non-null     object        \n",
      " 6   Localization                              19 non-null     object        \n",
      " 7   IDH1/2 status                             19 non-null     object        \n",
      " 8   MGMT status                               19 non-null     object        \n",
      " 9   ECOG                                      19 non-null     int64         \n",
      " 10  Nature of initial resection               19 non-null     object        \n",
      " 11  lymphocyte counts at V1 in G/L            19 non-null     float64       \n",
      " 12  cummulative dose preV1 steroid use in mg  19 non-null     int64         \n",
      " 13  steroid use after V1                      19 non-null     object        \n",
      " 14  cummulative steroid dose in mg            19 non-null     float64       \n",
      " 15   Cycles adj TMZ                           19 non-null     int64         \n",
      " 16  doses bevacizumab                         19 non-null     int64         \n",
      " 17  Best response                             19 non-null     object        \n",
      " 18  pseudo-progression (1=yes 0=no)           19 non-null     int64         \n",
      " 19  œdema                                     19 non-null     int64         \n",
      " 20  PP or œdema                               19 non-null     int64         \n",
      " 21  progression (1=yes 0=no)                  19 non-null     int64         \n",
      " 22  date of progression                       18 non-null     datetime64[ns]\n",
      " 23  months PFS                                19 non-null     int64         \n",
      " 24  deceased (1=yes 0=no)                     19 non-null     int64         \n",
      " 25  months OS                                 19 non-null     int64         \n",
      "dtypes: datetime64[ns](2), float64(2), int64(12), object(10)\n",
      "memory usage: 4.0+ KB\n"
     ]
    }
   ],
   "source": [
    "gdb1.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This dataset contains 26 fields of various datatypes. In one of the next steps this field names will be alligned with various controlled vocabularies and ontologies. For this to be succesful the fieldnames need to be as expressive as it can be. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_d718c_row3_col1, #T_d718c_row6_col1, #T_d718c_row10_col1, #T_d718c_row15_col1, #T_d718c_row18_col1, #T_d718c_row20_col1, #T_d718c_row21_col1, #T_d718c_row23_col1, #T_d718c_row24_col1, #T_d718c_row25_col1 {\n",
       "  background-color: yellow;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_d718c\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_d718c_level0_col0\" class=\"col_heading level0 col0\" >original_field</th>\n",
       "      <th id=\"T_d718c_level0_col1\" class=\"col_heading level0 col1\" >prepared_field</th>\n",
       "      <th id=\"T_d718c_level0_col2\" class=\"col_heading level0 col2\" >change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_d718c_row0_col0\" class=\"data row0 col0\" >ID Lab site</td>\n",
       "      <td id=\"T_d718c_row0_col1\" class=\"data row0 col1\" >ID Lab site</td>\n",
       "      <td id=\"T_d718c_row0_col2\" class=\"data row0 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_d718c_row1_col0\" class=\"data row1 col0\" >ID clinical site</td>\n",
       "      <td id=\"T_d718c_row1_col1\" class=\"data row1 col1\" >ID clinical site</td>\n",
       "      <td id=\"T_d718c_row1_col2\" class=\"data row1 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_d718c_row2_col0\" class=\"data row2 col0\" >DOB</td>\n",
       "      <td id=\"T_d718c_row2_col1\" class=\"data row2 col1\" >DOB</td>\n",
       "      <td id=\"T_d718c_row2_col2\" class=\"data row2 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_d718c_row3_col0\" class=\"data row3 col0\" >Age </td>\n",
       "      <td id=\"T_d718c_row3_col1\" class=\"data row3 col1\" >Age</td>\n",
       "      <td id=\"T_d718c_row3_col2\" class=\"data row3 col2\" >remove trailing spaces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_d718c_row4_col0\" class=\"data row4 col0\" >Gender</td>\n",
       "      <td id=\"T_d718c_row4_col1\" class=\"data row4 col1\" >Gender</td>\n",
       "      <td id=\"T_d718c_row4_col2\" class=\"data row4 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row5\" class=\"row_heading level0 row5\" >5</th>\n",
       "      <td id=\"T_d718c_row5_col0\" class=\"data row5 col0\" >Pathology</td>\n",
       "      <td id=\"T_d718c_row5_col1\" class=\"data row5 col1\" >Pathology</td>\n",
       "      <td id=\"T_d718c_row5_col2\" class=\"data row5 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row6\" class=\"row_heading level0 row6\" >6</th>\n",
       "      <td id=\"T_d718c_row6_col0\" class=\"data row6 col0\" >Localization </td>\n",
       "      <td id=\"T_d718c_row6_col1\" class=\"data row6 col1\" >Localization</td>\n",
       "      <td id=\"T_d718c_row6_col2\" class=\"data row6 col2\" >remove trailing spaces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row7\" class=\"row_heading level0 row7\" >7</th>\n",
       "      <td id=\"T_d718c_row7_col0\" class=\"data row7 col0\" >IDH1/2 status</td>\n",
       "      <td id=\"T_d718c_row7_col1\" class=\"data row7 col1\" >IDH1/2 status</td>\n",
       "      <td id=\"T_d718c_row7_col2\" class=\"data row7 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row8\" class=\"row_heading level0 row8\" >8</th>\n",
       "      <td id=\"T_d718c_row8_col0\" class=\"data row8 col0\" >MGMT status</td>\n",
       "      <td id=\"T_d718c_row8_col1\" class=\"data row8 col1\" >MGMT status</td>\n",
       "      <td id=\"T_d718c_row8_col2\" class=\"data row8 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row9\" class=\"row_heading level0 row9\" >9</th>\n",
       "      <td id=\"T_d718c_row9_col0\" class=\"data row9 col0\" >ECOG</td>\n",
       "      <td id=\"T_d718c_row9_col1\" class=\"data row9 col1\" >ECOG</td>\n",
       "      <td id=\"T_d718c_row9_col2\" class=\"data row9 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row10\" class=\"row_heading level0 row10\" >10</th>\n",
       "      <td id=\"T_d718c_row10_col0\" class=\"data row10 col0\" >Nature of initial resection </td>\n",
       "      <td id=\"T_d718c_row10_col1\" class=\"data row10 col1\" >Nature of initial resection</td>\n",
       "      <td id=\"T_d718c_row10_col2\" class=\"data row10 col2\" >remove trailing spaces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row11\" class=\"row_heading level0 row11\" >11</th>\n",
       "      <td id=\"T_d718c_row11_col0\" class=\"data row11 col0\" >lymphocyte counts at V1 in G/L</td>\n",
       "      <td id=\"T_d718c_row11_col1\" class=\"data row11 col1\" >lymphocyte counts at V1 in G/L</td>\n",
       "      <td id=\"T_d718c_row11_col2\" class=\"data row11 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row12\" class=\"row_heading level0 row12\" >12</th>\n",
       "      <td id=\"T_d718c_row12_col0\" class=\"data row12 col0\" >cummulative dose preV1 steroid use in mg</td>\n",
       "      <td id=\"T_d718c_row12_col1\" class=\"data row12 col1\" >cummulative dose preV1 steroid use in mg</td>\n",
       "      <td id=\"T_d718c_row12_col2\" class=\"data row12 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row13\" class=\"row_heading level0 row13\" >13</th>\n",
       "      <td id=\"T_d718c_row13_col0\" class=\"data row13 col0\" >steroid use after V1</td>\n",
       "      <td id=\"T_d718c_row13_col1\" class=\"data row13 col1\" >steroid use after V1</td>\n",
       "      <td id=\"T_d718c_row13_col2\" class=\"data row13 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row14\" class=\"row_heading level0 row14\" >14</th>\n",
       "      <td id=\"T_d718c_row14_col0\" class=\"data row14 col0\" >cummulative steroid dose in mg</td>\n",
       "      <td id=\"T_d718c_row14_col1\" class=\"data row14 col1\" >cummulative steroid dose in mg</td>\n",
       "      <td id=\"T_d718c_row14_col2\" class=\"data row14 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row15\" class=\"row_heading level0 row15\" >15</th>\n",
       "      <td id=\"T_d718c_row15_col0\" class=\"data row15 col0\" > Cycles adj TMZ</td>\n",
       "      <td id=\"T_d718c_row15_col1\" class=\"data row15 col1\" >Cycles adjuvant TMZ</td>\n",
       "      <td id=\"T_d718c_row15_col2\" class=\"data row15 col2\" >remove trailing spaces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row16\" class=\"row_heading level0 row16\" >16</th>\n",
       "      <td id=\"T_d718c_row16_col0\" class=\"data row16 col0\" >doses bevacizumab</td>\n",
       "      <td id=\"T_d718c_row16_col1\" class=\"data row16 col1\" >doses bevacizumab</td>\n",
       "      <td id=\"T_d718c_row16_col2\" class=\"data row16 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row17\" class=\"row_heading level0 row17\" >17</th>\n",
       "      <td id=\"T_d718c_row17_col0\" class=\"data row17 col0\" >Best response</td>\n",
       "      <td id=\"T_d718c_row17_col1\" class=\"data row17 col1\" >Best response</td>\n",
       "      <td id=\"T_d718c_row17_col2\" class=\"data row17 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row18\" class=\"row_heading level0 row18\" >18</th>\n",
       "      <td id=\"T_d718c_row18_col0\" class=\"data row18 col0\" >pseudo-progression (1=yes 0=no)</td>\n",
       "      <td id=\"T_d718c_row18_col1\" class=\"data row18 col1\" >pseudo-progression </td>\n",
       "      <td id=\"T_d718c_row18_col2\" class=\"data row18 col2\" >removed choices</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row19\" class=\"row_heading level0 row19\" >19</th>\n",
       "      <td id=\"T_d718c_row19_col0\" class=\"data row19 col0\" >œdema</td>\n",
       "      <td id=\"T_d718c_row19_col1\" class=\"data row19 col1\" >œdema</td>\n",
       "      <td id=\"T_d718c_row19_col2\" class=\"data row19 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row20\" class=\"row_heading level0 row20\" >20</th>\n",
       "      <td id=\"T_d718c_row20_col0\" class=\"data row20 col0\" >PP or œdema</td>\n",
       "      <td id=\"T_d718c_row20_col1\" class=\"data row20 col1\" >pseudo-progression or œdema</td>\n",
       "      <td id=\"T_d718c_row20_col2\" class=\"data row20 col2\" >resolved abbreviations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row21\" class=\"row_heading level0 row21\" >21</th>\n",
       "      <td id=\"T_d718c_row21_col0\" class=\"data row21 col0\" >progression (1=yes 0=no)</td>\n",
       "      <td id=\"T_d718c_row21_col1\" class=\"data row21 col1\" >progression </td>\n",
       "      <td id=\"T_d718c_row21_col2\" class=\"data row21 col2\" >removed choices</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row22\" class=\"row_heading level0 row22\" >22</th>\n",
       "      <td id=\"T_d718c_row22_col0\" class=\"data row22 col0\" >date of progression</td>\n",
       "      <td id=\"T_d718c_row22_col1\" class=\"data row22 col1\" >date of progression</td>\n",
       "      <td id=\"T_d718c_row22_col2\" class=\"data row22 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row23\" class=\"row_heading level0 row23\" >23</th>\n",
       "      <td id=\"T_d718c_row23_col0\" class=\"data row23 col0\" >months PFS</td>\n",
       "      <td id=\"T_d718c_row23_col1\" class=\"data row23 col1\" >months progression-free survival</td>\n",
       "      <td id=\"T_d718c_row23_col2\" class=\"data row23 col2\" >resolved abbreviations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row24\" class=\"row_heading level0 row24\" >24</th>\n",
       "      <td id=\"T_d718c_row24_col0\" class=\"data row24 col0\" >deceased (1=yes 0=no)</td>\n",
       "      <td id=\"T_d718c_row24_col1\" class=\"data row24 col1\" >deceased </td>\n",
       "      <td id=\"T_d718c_row24_col2\" class=\"data row24 col2\" >removed choices</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d718c_level0_row25\" class=\"row_heading level0 row25\" >25</th>\n",
       "      <td id=\"T_d718c_row25_col0\" class=\"data row25 col0\" >months OS</td>\n",
       "      <td id=\"T_d718c_row25_col1\" class=\"data row25 col1\" >months overall survival</td>\n",
       "      <td id=\"T_d718c_row25_col2\" class=\"data row25 col2\" >resolved abbreviations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x13b6bb850>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def show_difference(row):\n",
    "    highlight = 'background-color: yellow;'\n",
    "    default = ''\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        return [default, highlight]\n",
    "    else:\n",
    "        return [default, default]\n",
    "    \n",
    "def strip_check(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        row[\"change\"] = \"remove trailing spaces\"\n",
    "        \n",
    "def abbreviations(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        if row[\"change\"] == \"\":\n",
    "            row[\"change\"] = \"resolved abbreviations\"\n",
    "            \n",
    "def removechoices(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        if row[\"change\"] == \"\":\n",
    "            row[\"change\"] = \"removed choices\"\n",
    "\n",
    "df = pd.DataFrame(columns=[\"original_field\", \"prepared_field\", \"change\"])\n",
    "for column in gdb1.columns: \n",
    "    df.loc[len(df.index)] = [column, column.strip(), \"\"]\n",
    "    \n",
    "df.apply(strip_check, axis=1)\n",
    "    \n",
    "df['prepared_field'] = df['prepared_field'].str.replace('adj','adjuvant')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('months OS','months overall survival')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('PP','pseudo-progression')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('PFS','progression-free survival')\n",
    "\n",
    "df.apply(abbreviations, axis=1)\n",
    "         \n",
    "df['prepared_field'] = df['prepared_field'].str.replace('(1=yes 0=no)','', regex=False)\n",
    "df.apply(removechoices, axis=1)\n",
    " \n",
    "df.style.set_properties(**{'text-align': 'left'})\n",
    "df.style.apply(show_difference, subset=['original_field', 'prepared_field'], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataset 2 on Glioblastoma\n",
    "The second dataset is a provided excel file (dt_Bayraktar_ClinicalMetadata_mock_2022.07.21.xlsx) ## TODO add a link. It contains synthetic clinical data on patients with glioblastoma."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "gdb2 = pd.read_excel(\"data/dt_Bayraktar_ClinicalMetadata_mock_2022.07.21.xlsx\")\n",
    "gdb2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "gdb2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "def show_difference(row):\n",
    "    highlight = 'background-color: yellow;'\n",
    "    default = ''\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        return [default, highlight]\n",
    "    else:\n",
    "        return [default, default]\n",
    "    \n",
    "def strip_check(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        row[\"change\"] = \"remove trailing spaces\"\n",
    "        \n",
    "def abbreviations(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        if row[\"change\"] == \"\":\n",
    "            row[\"change\"] = \"resolved abbreviations\"\n",
    "            \n",
    "def removechoices(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        if row[\"change\"] == \"\":\n",
    "            row[\"change\"] = \"removed choices\"\n",
    "\n",
    "df = pd.DataFrame(columns=[\"original_field\", \"prepared_field\", \"change\"])\n",
    "for column in gdb2.columns: \n",
    "    df.loc[len(df.index)] = [column, column.strip(), \"\"]\n",
    "    \n",
    "df.apply(strip_check, axis=1)\n",
    "    \n",
    "df['prepared_field'] = df['prepared_field'].str.replace('adj','adjuvant')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('months OS','months overall survival')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('PP','pseudo-progression')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('PFS','progression-free survival')\n",
    "\n",
    "df.apply(abbreviations, axis=1)\n",
    "         \n",
    "df['prepared_field'] = df['prepared_field'].str.replace('(1=yes 0=no)','', regex=False)\n",
    "df.apply(removechoices, axis=1)\n",
    " \n",
    "df.style.set_properties(**{'text-align': 'left'})\n",
    "df.style.apply(show_difference, subset=['original_field', 'prepared_field'], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the next chapter the clarified field names will be used to identify terms in applicable controlled vocabularies ontologies to drive the linked data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>MPBTP (Sequencing Data)</th>\n",
       "      <th># of blocks</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age (optional)</th>\n",
       "      <th>Tumour Spatial Information</th>\n",
       "      <th>Pathology - pre-surgical</th>\n",
       "      <th>Histopathological Assignment</th>\n",
       "      <th>Histopathology - Microscopic Description</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>Methylation Status (&gt;10% = METHYLATED)</th>\n",
       "      <th>Other Variants</th>\n",
       "      <th>Covid test</th>\n",
       "      <th>Other comments</th>\n",
       "      <th>Spatial Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Patient A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>M</td>\n",
       "      <td>54</td>\n",
       "      <td>Right Temporal Lobe</td>\n",
       "      <td>GBM</td>\n",
       "      <td>Glioblastoma, IDH-wiltype (CNS WHO Grade 4)</td>\n",
       "      <td>Cellular glial neoplasm with elongated nuclei ...</td>\n",
       "      <td>IDH-WT (IHC and Sequencing)\\nATRX - Retained</td>\n",
       "      <td>MGMT Promotor Methylation - 2% - ABSENT</td>\n",
       "      <td>EGFR Amplification;\\nCDKN2A Deletion;\\nPTEN De...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>A1 has a duplicate block</td>\n",
       "      <td>Wasn’t able to get stealth location informatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Patient B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>F</td>\n",
       "      <td>82</td>\n",
       "      <td>Right Temporal Lobe</td>\n",
       "      <td>GBM</td>\n",
       "      <td>Glioblastoma, IDH-wiltype (CNS WHO Grade 4)</td>\n",
       "      <td>Pleomorphic astrocytes with numerous gemistocy...</td>\n",
       "      <td>IDH-WT (IHC and Sequencing)\\nATRX - Retained</td>\n",
       "      <td>MGMT Promotor Methylation - 5% - ABSENT</td>\n",
       "      <td>CDKN2A Deletion; \\nPTEN Deletion</td>\n",
       "      <td>Negative</td>\n",
       "      <td>-</td>\n",
       "      <td>Anterior, Lateral, Posterior and Medial sample...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Patient C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>F</td>\n",
       "      <td>41</td>\n",
       "      <td>Right Frontal Lobe</td>\n",
       "      <td>GBM</td>\n",
       "      <td>Oligodendroglioma, IDH-mut and 1p/19q co-delet...</td>\n",
       "      <td>Diffusely infiltrative glial neoplasm with mod...</td>\n",
       "      <td>IDH1-mut R132H (IHC and Sequencing)\\nATRX - Re...</td>\n",
       "      <td>No tested</td>\n",
       "      <td>1p/19q Co-deletion; \\nTP53 P152L</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Sampled by Richard</td>\n",
       "      <td>Superior, Anterior, Deep and Inferior samples ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample ID  MPBTP (Sequencing Data)  # of blocks Sex  Age (optional)  \\\n",
       "0  Patient A                      NaN            5   M              54   \n",
       "1  Patient B                      NaN            4   F              82   \n",
       "2  Patient C                      NaN            4   F              41   \n",
       "\n",
       "  Tumour Spatial Information  Pathology - pre-surgical  \\\n",
       "0         Right Temporal Lobe                     GBM    \n",
       "1         Right Temporal Lobe                     GBM    \n",
       "2          Right Frontal Lobe                     GBM    \n",
       "\n",
       "                        Histopathological Assignment  \\\n",
       "0        Glioblastoma, IDH-wiltype (CNS WHO Grade 4)   \n",
       "1        Glioblastoma, IDH-wiltype (CNS WHO Grade 4)   \n",
       "2  Oligodendroglioma, IDH-mut and 1p/19q co-delet...   \n",
       "\n",
       "            Histopathology - Microscopic Description  \\\n",
       "0  Cellular glial neoplasm with elongated nuclei ...   \n",
       "1  Pleomorphic astrocytes with numerous gemistocy...   \n",
       "2  Diffusely infiltrative glial neoplasm with mod...   \n",
       "\n",
       "                                           Mutations  \\\n",
       "0       IDH-WT (IHC and Sequencing)\\nATRX - Retained   \n",
       "1       IDH-WT (IHC and Sequencing)\\nATRX - Retained   \n",
       "2  IDH1-mut R132H (IHC and Sequencing)\\nATRX - Re...   \n",
       "\n",
       "    Methylation Status (>10% = METHYLATED)  \\\n",
       "0  MGMT Promotor Methylation - 2% - ABSENT   \n",
       "1  MGMT Promotor Methylation - 5% - ABSENT   \n",
       "2                                No tested   \n",
       "\n",
       "                                      Other Variants Covid test  \\\n",
       "0  EGFR Amplification;\\nCDKN2A Deletion;\\nPTEN De...   Negative   \n",
       "1                   CDKN2A Deletion; \\nPTEN Deletion   Negative   \n",
       "2                   1p/19q Co-deletion; \\nTP53 P152L   Negative   \n",
       "\n",
       "             Other comments                                   Spatial Location  \n",
       "0  A1 has a duplicate block  Wasn’t able to get stealth location informatio...  \n",
       "1                         -  Anterior, Lateral, Posterior and Medial sample...  \n",
       "2        Sampled by Richard  Superior, Anterior, Deep and Inferior samples ...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdb2 = pd.read_excel(\"data/dt_Bayraktar_ClinicalMetadata_mock_2022.07.21.xlsx\")\n",
    "gdb2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 3 entries, 0 to 2\n",
      "Data columns (total 15 columns):\n",
      " #   Column                                    Non-Null Count  Dtype  \n",
      "---  ------                                    --------------  -----  \n",
      " 0   Sample ID                                 3 non-null      object \n",
      " 1   MPBTP (Sequencing Data)                   0 non-null      float64\n",
      " 2   # of blocks                               3 non-null      int64  \n",
      " 3   Sex                                       3 non-null      object \n",
      " 4   Age (optional)                            3 non-null      int64  \n",
      " 5   Tumour Spatial Information                3 non-null      object \n",
      " 6   Pathology - pre-surgical                  3 non-null      object \n",
      " 7   Histopathological Assignment              3 non-null      object \n",
      " 8   Histopathology - Microscopic Description  3 non-null      object \n",
      " 9   Mutations                                 3 non-null      object \n",
      " 10  Methylation Status (>10% = METHYLATED)    3 non-null      object \n",
      " 11  Other Variants                            3 non-null      object \n",
      " 12  Covid test                                3 non-null      object \n",
      " 13  Other comments                            3 non-null      object \n",
      " 14  Spatial Location                          3 non-null      object \n",
      "dtypes: float64(1), int64(2), object(12)\n",
      "memory usage: 488.0+ bytes\n"
     ]
    }
   ],
   "source": [
    "gdb2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_4c26c_row4_col1, #T_4c26c_row5_col1 {\n",
       "  background-color: yellow;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_4c26c\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_4c26c_level0_col0\" class=\"col_heading level0 col0\" >original_field</th>\n",
       "      <th id=\"T_4c26c_level0_col1\" class=\"col_heading level0 col1\" >prepared_field</th>\n",
       "      <th id=\"T_4c26c_level0_col2\" class=\"col_heading level0 col2\" >change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_4c26c_row0_col0\" class=\"data row0 col0\" >Sample ID</td>\n",
       "      <td id=\"T_4c26c_row0_col1\" class=\"data row0 col1\" >Sample ID</td>\n",
       "      <td id=\"T_4c26c_row0_col2\" class=\"data row0 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_4c26c_row1_col0\" class=\"data row1 col0\" >MPBTP (Sequencing Data)</td>\n",
       "      <td id=\"T_4c26c_row1_col1\" class=\"data row1 col1\" >MPBTP (Sequencing Data)</td>\n",
       "      <td id=\"T_4c26c_row1_col2\" class=\"data row1 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_4c26c_row2_col0\" class=\"data row2 col0\" ># of blocks</td>\n",
       "      <td id=\"T_4c26c_row2_col1\" class=\"data row2 col1\" ># of blocks</td>\n",
       "      <td id=\"T_4c26c_row2_col2\" class=\"data row2 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_4c26c_row3_col0\" class=\"data row3 col0\" >Sex</td>\n",
       "      <td id=\"T_4c26c_row3_col1\" class=\"data row3 col1\" >Sex</td>\n",
       "      <td id=\"T_4c26c_row3_col2\" class=\"data row3 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_4c26c_row4_col0\" class=\"data row4 col0\" >Age (optional)</td>\n",
       "      <td id=\"T_4c26c_row4_col1\" class=\"data row4 col1\" >Age </td>\n",
       "      <td id=\"T_4c26c_row4_col2\" class=\"data row4 col2\" >removed conditionals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row5\" class=\"row_heading level0 row5\" >5</th>\n",
       "      <td id=\"T_4c26c_row5_col0\" class=\"data row5 col0\" >Tumour Spatial Information </td>\n",
       "      <td id=\"T_4c26c_row5_col1\" class=\"data row5 col1\" >Tumour Spatial Information</td>\n",
       "      <td id=\"T_4c26c_row5_col2\" class=\"data row5 col2\" >remove trailing spaces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row6\" class=\"row_heading level0 row6\" >6</th>\n",
       "      <td id=\"T_4c26c_row6_col0\" class=\"data row6 col0\" >Pathology - pre-surgical</td>\n",
       "      <td id=\"T_4c26c_row6_col1\" class=\"data row6 col1\" >Pathology - pre-surgical</td>\n",
       "      <td id=\"T_4c26c_row6_col2\" class=\"data row6 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row7\" class=\"row_heading level0 row7\" >7</th>\n",
       "      <td id=\"T_4c26c_row7_col0\" class=\"data row7 col0\" >Histopathological Assignment</td>\n",
       "      <td id=\"T_4c26c_row7_col1\" class=\"data row7 col1\" >Histopathological Assignment</td>\n",
       "      <td id=\"T_4c26c_row7_col2\" class=\"data row7 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row8\" class=\"row_heading level0 row8\" >8</th>\n",
       "      <td id=\"T_4c26c_row8_col0\" class=\"data row8 col0\" >Histopathology - Microscopic Description</td>\n",
       "      <td id=\"T_4c26c_row8_col1\" class=\"data row8 col1\" >Histopathology - Microscopic Description</td>\n",
       "      <td id=\"T_4c26c_row8_col2\" class=\"data row8 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row9\" class=\"row_heading level0 row9\" >9</th>\n",
       "      <td id=\"T_4c26c_row9_col0\" class=\"data row9 col0\" >Mutations</td>\n",
       "      <td id=\"T_4c26c_row9_col1\" class=\"data row9 col1\" >Mutations</td>\n",
       "      <td id=\"T_4c26c_row9_col2\" class=\"data row9 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row10\" class=\"row_heading level0 row10\" >10</th>\n",
       "      <td id=\"T_4c26c_row10_col0\" class=\"data row10 col0\" >Methylation Status (>10% = METHYLATED)</td>\n",
       "      <td id=\"T_4c26c_row10_col1\" class=\"data row10 col1\" >Methylation Status (>10% = METHYLATED)</td>\n",
       "      <td id=\"T_4c26c_row10_col2\" class=\"data row10 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row11\" class=\"row_heading level0 row11\" >11</th>\n",
       "      <td id=\"T_4c26c_row11_col0\" class=\"data row11 col0\" >Other Variants</td>\n",
       "      <td id=\"T_4c26c_row11_col1\" class=\"data row11 col1\" >Other Variants</td>\n",
       "      <td id=\"T_4c26c_row11_col2\" class=\"data row11 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row12\" class=\"row_heading level0 row12\" >12</th>\n",
       "      <td id=\"T_4c26c_row12_col0\" class=\"data row12 col0\" >Covid test</td>\n",
       "      <td id=\"T_4c26c_row12_col1\" class=\"data row12 col1\" >Covid test</td>\n",
       "      <td id=\"T_4c26c_row12_col2\" class=\"data row12 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row13\" class=\"row_heading level0 row13\" >13</th>\n",
       "      <td id=\"T_4c26c_row13_col0\" class=\"data row13 col0\" >Other comments</td>\n",
       "      <td id=\"T_4c26c_row13_col1\" class=\"data row13 col1\" >Other comments</td>\n",
       "      <td id=\"T_4c26c_row13_col2\" class=\"data row13 col2\" ></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_4c26c_level0_row14\" class=\"row_heading level0 row14\" >14</th>\n",
       "      <td id=\"T_4c26c_row14_col0\" class=\"data row14 col0\" >Spatial Location</td>\n",
       "      <td id=\"T_4c26c_row14_col1\" class=\"data row14 col1\" >Spatial Location</td>\n",
       "      <td id=\"T_4c26c_row14_col2\" class=\"data row14 col2\" ></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x12dc97bb0>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def show_difference(row):\n",
    "    highlight = 'background-color: yellow;'\n",
    "    default = ''\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        return [default, highlight]\n",
    "    else:\n",
    "        return [default, default]\n",
    "    \n",
    "def strip_check(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        row[\"change\"] = \"remove trailing spaces\"\n",
    "        \n",
    "def abbreviations(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        if row[\"change\"] == \"\":\n",
    "            row[\"change\"] = \"resolved abbreviations\"\n",
    "            \n",
    "def removechoices(row):\n",
    "    if row['original_field'] != row['prepared_field']:\n",
    "        if row[\"change\"] == \"\":\n",
    "            row[\"change\"] = \"removed conditionals\"\n",
    "\n",
    "df = pd.DataFrame(columns=[\"original_field\", \"prepared_field\", \"change\"])\n",
    "for column in gdb2.columns: \n",
    "    df.loc[len(df.index)] = [column, column.strip(), \"\"]\n",
    "    \n",
    "df.apply(strip_check, axis=1)\n",
    "    \n",
    "df['prepared_field'] = df['prepared_field'].str.replace('adj','adjuvant')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('months OS','months overall survival')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('PP','pseudo-progression')\n",
    "df['prepared_field'] = df['prepared_field'].str.replace('PFS','progression-free survival')\n",
    "\n",
    "df.apply(abbreviations, axis=1)\n",
    "         \n",
    "df['prepared_field'] = df['prepared_field'].str.replace('(optional)','', regex=False)\n",
    "df.apply(removechoices, axis=1)\n",
    " \n",
    "df.style.set_properties(**{'text-align': 'left'})\n",
    "df.style.apply(show_difference, subset=['original_field', 'prepared_field'], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the next chapter the clarified field names will be used to identify terms in applicable controlled vocabularies ontologies to drive the linked data. "
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
